lib/codegen/src/regalloc/spilling.rs

```diff
@@ -335,6 +344,29 @@ impl<'a> Context<'a> {
                 self.reg_uses.push(reguse);
             }
         }
+
+        // Similarly, for return instructions, collect uses of ABI-defined
+        // return values.
+        if self.cur.func.dfg[inst].opcode().is_return() {
+            for (ret_idx, (ret, &arg)) in
+                self.cur.func.signature.returns.iter().zip(args).enumerate()
+            {
+                let idx = constraints.ins.len() + ret_idx;
```

Comment:
This is a bit unfortunate, because if I understand correctly, this breaks encapsulation: the layout of the arguments vector of the return instruction in particular is hereby exposed. (I've seen this index is used in `process_reg_uses` as an index into the arguments vector) Maybe there's a way to encapsulate it a bit better? (could be done as a follow-up)

Reply:
The loop above this is iterating over the instruction args zipped with `constraints.ins.iter()`, so it already knows about the layout of the arguements vector. By starting at `constraints.ins.len()`, we're effectively just picking up the arguements where the previous loop left off.

More generally, we make these assumptions in a lot of places:
 - fixed arguments are described by the opcode constraints, variable arguments are not
 - variable arguments appear after the fixed arguments

and I think these are reasonable, though I'm open to discussing other approaches.

I have added an assert to the PR to at least check that the `return` has the number of return values the signature says it should, which should help catch any cases where things get out of sync.

Reply:
OK, I wasn't aware of this but this comment sheds light on this invariant, so that's good enough to me. Thanks!